iset.mm

```diff
@@ -38820,6 +38826,22 @@ class of all ordinal numbers to be a set (so instead it is a proper
       WBVPVGEVNVFVAUKUNULVSVJBVNVFVICVNVFUOUMUPFUQURAGCBVFVLUSUT $.
   $}
 
+  ${
+    $d ph x y $.
+    dcextest.ex $e |- DECID { x | ph } e. _V $.
+    $( If the proposition ` A e. _V ` is decidable for any class, any negated
+       proposition is decidable.  Here we set ` A ` to be ` { x | ph } ` (since
+       we have no way to express "for any class A", we have to write the
+       theorem with a particular ` A ` ).  (Contributed by Jim Kingdon,
+       3-Jul-2022.) $)
+    dcextest $p |- ( -. ph \/ -. -. ph ) $=
+      ( vy cab cvv wcel wn wo wdc exmiddc ax-mp elirr weq df-v wb equid pm5.1im
```

Comment:
why is ~elirr used here, although its usage is discouraged?

Reply:
It is only discouraged for ordinals, as described at http://us.metamath.org/ileuni/elirr.html we'd like to use http://us.metamath.org/ileuni/ordirr.html then.

Reply:
@jkingdon It is discouraged everywhere, because it depends on the foundation axiom and you can usually prove equivalent results without it. How is it being used in this proof?

Reply:
> @jkingdon It is discouraged everywhere, because it depends on the foundation axiom and you can usually prove equivalent results without it.

LOL, I won't belabor the point of who put in the discouraged tag and based on what reasoning, but let's talk about how foundation is being used here rather than procedure.

> How is it being used in this proof?

Same way it is being used in http://us.metamath.org/ileuni/ruv.html and http://us.metamath.org/mpeuni/ruv.html

I didn't see a way to adapt the http://us.metamath.org/ileuni/ru.html proof or another one I thought of to this situation, but I suppose this probably can be proved without foundation. Maybe I could get a bit of help from people who want to reduce the use of foundation (either in terms of this particular proof, or the 14 existing usages)?

Reply:
Oops, I didn't realize it had an actual discouraged tag. I meant it in the colloquial sense. It's possible that the tag was added to avoid introducing axioms? But we already have the `/no_new_axioms` flag for that. :shrug: 

The way you can avoid it here is by using vprc instead, which is proved not by showing that the russell class is *equal* to V but rather that it is a *subset* of V and hence if V is a set then the russell class is too. That is, under the assumption `ph`, `{ x | ph } = _V e/ _V`, so by contrapositive `{ x | ph } e. _V` implies `-. ph`.

Reply:
> It's possible that the tag was added to avoid introducing axioms? But we already have the `/no_new_axioms` flag for that. shrug

`/no_new_axioms` won't (currently) help (in fact, I added the discouraged tag when I noticed the minimizer removing the `ordirr` which I had carefully inserted), To explain I'll just quote the comment of http://us.metamath.org/ileuni/elirr.html :

> The reason that this theorem is marked as discouraged is a bit subtle. If we wanted to reduce usage of [ax-setind](http://us.metamath.org/ileuni/ax-setind.html) 4309, we could redefine Ord 𝐴 ([df-iord](http://us.metamath.org/ileuni/df-iord.html) 4150) to also require E Fr 𝐴 ([df-frind](http://us.metamath.org/ileuni/df-frind.html) 4116) and in that case any theorem related to irreflexivity of ordinals could use [ordirr](http://us.metamath.org/ileuni/ordirr.html) 4314 (which under that definition would presumably not need [ax-setind](http://us.metamath.org/ileuni/ax-setind.html) 4309 to prove it). But since ordinals have not yet been defined that way, we cannot rely on the "don't add additional axiom use" feature of the minimizer to get theorems to use [ordirr](http://us.metamath.org/ileuni/ordirr.html) 4314. To encourage [ordirr](http://us.metamath.org/ileuni/ordirr.html) 4314 when possible, we mark this theorem as discouraged.

_going back to the message I'm responding to:_
> The way you can avoid it here is by using vprc instead

Ah, cool. That's easier than I had feared. Done.

> By the way, on the subject of snex and constant classes, one other counterintuitive fact I noticed is that `{ { x | ph } } e. _V`, because it is unconditionally a subset of `1o`, because any element of the singleton `y = { x | ph }` implies that `-. ph` and hence `y = (/)`.

This is an odd little theorem, and I'll agree with counterintuitive. Which I guess is an argument for formalizing this proof. Done.

Reply:
> I don't understand why you don't use "`DECID -. ph`" for the conclusion.

Well, I could offer an impassioned defence about how overusing definitions increases the cognitive load especially for people new to this material, who are least able to bear it. However, I think that might fall kind of flat in a case where the hypothesis of the very same theorem uses `DECID` so, OK, there seems to be little reason to avoid it here. Changed.

> For the comment, I would write:
> 
> > If it is decidable whether `{ x | ph }` is a set, then `-. ph` is decidable (where `x` does not occur in `ph`). From this fact, we can deduce (outside the formal system, since we cannot quantify over classes) that if it is decidable whether any class is a set, then "weak excluded middle" (that is, any negated proposition `-. ph` is decidable) holds.

I'm not in love with either that wording or the one I had, but I guess this looks good (for example, defines "weak excluded middle" rather than assuming people will know/remember what that is), so I'l use it. Wording adopted.

Reply:
> overusing definitions increases the cognitive load especially for people new to this material, who are least able to bear it

I think this is an important concern. But to me this is more of an argument to introduce a definition only if it's "really" needed, but once it is introduced, we might as well use it without being coy (there is even an argument to say that if we use it only in certain cases, then it will remain obscure, whereas if it's systematically used, the reader will become familiar with it).

> I'm not in love with either that wording or the one I had, but I guess this looks good (for example, defines "weak excluded middle" rather than assuming people will know/remember what that is), so I'l use it. Wording adopted.

Sometimes my English grammar is a bit rusty, so please correct it when necessary. As for the comment itself, I prefer to first state precisely what the formal theorem proves, and then digressing about what metatheorem we can deduce, rather than the opposite ("we want to prove that, but we can't within the formal system, so we prove that instead").  This of course is only valid for "small units" (in an article, or on the home page for ileuni, it is generally best to begin with an informal explanation).

Reply:
> But to me this is more of an argument to introduce a definition only if it's "really" needed,

To give an example, the definition of "testable" is not really needed, since `TEST ph` is simply defined as `DECID -. ph` (so even when you substitute a huge expression for `ph`), there is no gain in having it (which is why it was recently removed).  Not unrelatedly, the meaning of "testable" is certainly much less known than that of "decidable". (I just realized that the "test" in the label "dcextest" was referring to testability of ph).